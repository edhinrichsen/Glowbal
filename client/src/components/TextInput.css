.text-input-display-none {
    /* display: none; */
    width: 0;
    height: 0;
    overflow: scroll;
    
}

.text-input {
    /* background-color: blueviolet; */
    cursor: default;
    /* margin: -5px; */
    width: 100%;
    /* height: fit-content; */
    border-radius: var(--border-radius-secondary);
    transition: box-shadow 0.3s;
    /* background-color: rosybrown; */
}


.text-input-editable:hover {
    box-shadow: inset var(--drop-shadow);
    cursor: text;
    
}

.text-input-focus {
    box-shadow: inset var(--drop-shadow);
    
}

.text-input input {
    text-align: inherit;
    font-weight: inherit;
    font-size: inherit;
    font-style: inherit;
    font-family: inherit;
    border: none;
    outline: none;
    padding: 0;
    margin: 0;
    width: 100%;
    background-color: rgba(0, 0, 0, 0);
}