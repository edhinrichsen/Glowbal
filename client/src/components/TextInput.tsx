import React, { useRef, useState } from 'react';

import './TextInput.css';

type TextInputProps = {
  text?: string;
  padding?: string;
  radius?: string;
  editable?: boolean;
  multiline?: boolean;
  maxLen?: number;
  onChange: Function;
};

function TextInput(props: TextInputProps) {
  function handleChange(event: { target: { value: string } }) {
    if (props.maxLen) {
      if (event.target.value.length < props.maxLen) {
        setText(event.target.value);
      } else {
        setText(event.target.value.substr(0, props.maxLen));
      }
    } else {
      setText(event.target.value);
    }
  }
  const [text, setText] = useState(props.text);
  const [editing, setEditing] = useState(false);
  const inputRef = useRef<any>(null);
  const displayTextRef = useRef<any>(null);

  const finish = () => {
    setEditing(false);
    props.onChange(text);
  };
  return (
    <div
      style={{ padding: props.padding, borderRadius: props.radius }}
      className={
        editing
          ? 'text-input text-input-focus'
          : props.editable
          ? 'text-input text-input-editable'
          : 'text-input'
      }
      onClick={() => {
        if (props.editable) {
          setText(props.text);
          setEditing(true);
          if (inputRef.current !== null && displayTextRef.current !== null) {
            setTimeout(() => inputRef.current.focus(), 50);

            let width = displayTextRef.current.offsetWidth;
            if (width > 0) {
              inputRef.current.style.width = width + 'px';
            }

            let height = displayTextRef.current.offsetHeight;
            if (height > 0) {
              inputRef.current.style.height = height + 'px';
            }
          }
        }
      }}
    >
      <div
        ref={displayTextRef}
        className={editing ? 'text-input-display-none' : ''}
      >
        {props.text}
      </div>
      {props.multiline ? (
        <div className={editing ? '' : 'text-input-display-none'}>
          <textarea
            className="container-scroll"
            ref={inputRef}
            onBlur={() => finish()}
            value={text}
            onChange={handleChange}
          />
        </div>
      ) : (
        <div className={editing ? '' : 'text-input-display-none'}>
          <input
            ref={inputRef}
            onBlur={() => finish()}
            type="text"
            value={text}
            onChange={handleChange}
          />
        </div>
      )}
    </div>
  );
}

export default TextInput;
