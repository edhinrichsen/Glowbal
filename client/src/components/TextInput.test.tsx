import React from 'react';
import { shallow } from 'enzyme';
import TextInput from './TextInput';

describe('Testing TextInput', () => {
  it('renders without crashing', () => {
    const component = shallow(<TextInput onChange={() => {}} />);
    expect(component).toMatchSnapshot();
    component.unmount();
  });
});
