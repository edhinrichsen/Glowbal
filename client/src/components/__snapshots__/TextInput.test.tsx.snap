// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing TextInput renders without crashing 1`] = `
<div
  className="text-input"
  onClick={[Function]}
  style={
    Object {
      "borderRadius": undefined,
      "padding": undefined,
    }
  }
>
  <div
    className=""
  >
    <span
      className="placeholder"
    />
  </div>
  <div
    className="text-input-display-none"
  >
    <input
      onBlur={[Function]}
      onChange={[Function]}
      placeholder=""
      type="text"
    />
  </div>
</div>
`;
